src/lplib/cmap.ml

```diff
@@ -0,0 +1,174 @@
+with start = finish) inside an interval.
+
+In any other case, overlapping intervals
+can't be compared and will throw an error.
+
+An interval is considered "smaller" than another
+if all its points are smaller that the points
+of the other (i.e its finishing point is before
+the starting point of the other). *)
+module Range : RangeType =
+  struct
+    type point = { l : int; c : int }
+    let make_point l c = { l = l; c = c}
+
+    type t = point * point
+    type interval = t
+    let make_interval x y = 
+      assert (x <= y);
+      (x, y)
+
+    let line pt = pt.l
+    let column pt = pt .c
+
+    let interval_start = fst
+    let interval_end = snd
+
+    type cmp = Before | In | After
+
+    let in_range pos (pos1, pos2) = 
+      
+      (*Comparison operators work as lexicographic comparison
+      on points (meaning l is compared, then c is compared). *)
+      if pos < pos1
+      then Before
+
+      else if pos > pos2
+      then After
+
+      else In
+    
+    let compare (s1, f1) (s2, f2) =
+
+      (*Two intervals are considered equal if one is included in the other. *)
+      if (s1 >= s2 && f1 <= f2) || (s1 <= s2 && f1 >= f2)
+      then 0
+
+      else if f1 <= s2
+      then -1
+
+      else if s1 >= f2
+      then 1
+
+      else
+        failwith "Intervals overlap, no inclusion between the two"
+
+    let point_to_string pt = "Line : " ^ string_of_int pt.l
+      ^ "; Column : " ^ string_of_int pt.c ^ "\n"
+
+    let interval_to_string interv =
+      let s, f = (interval_start interv), (interval_end interv) in
+      "From :\n" ^ point_to_string s ^ "To :\n" ^ point_to_string f
+  end
+
+(*A structure which maps a position of the cursor to the corresponding token
+in a given text editor with a .lp file open.
+t is the type of maps while 'a is the type of tokens/values stored
+in the map. *)
+module type RangeMapType = 
+  sig
+    type 'a t
+    type interval
+    type point
+
+    (*Given a cursor position, returns the corresponding token. *)
+    val find : point -> 'a t -> (interval * 'a) option
+
+    val empty : 'a t
+
+    (*Adds the key (inteval) for a token and the token in a map. *)
+    (*Requires all keys to be non overlapping intervals. *)
+    (*/!\ Does not ensure proper functionning if the added keys are overlapping
+    intervals, e.g might change a previously added (key, element) couple
+    or throw an error. *)
+    val add : interval -> 'a -> 'a t -> 'a t
+
+    val to_string : ('a -> string) -> 'a t -> string
+  end
+
+(*The functor for cursor maps. *)
+module MakeRangeMap (Range : RangeType) =
+  struct
+    (*A map of which keys are intervals. *)
+    module RangeMap = Map.Make(Range)
+  
+    (*Now we need to transform the map so that :
+      - the keys for "add" are intervals
+      - the keys for "find" are points.
+    *)
+    type 'a t = (Range.interval * 'a) RangeMap.t
+    type interval = Range.interval
+    type point = Range.point
```

Comment:
I need types interval and point in the signature for range maps, which can't be made a functor since it is a signature, and with the recommendation of Emilio I made range maps generic so that any implementation of Range can be used. Given these criteria, I don't see any other way to do this.

Reply:
Solved.